torchvision/models/detection/retinanet.py

```diff
@@ -0,0 +1,572 @@
+
+    @torch.jit.unused
+    def eager_outputs(self, losses, detections):
+        # type: (Dict[str, Tensor], List[Dict[str, Tensor]]) -> Tuple[Dict[str, Tensor], List[Dict[str, Tensor]]]
+        if self.training:
+            return losses
+
+        return detections
+
+    def compute_loss(self, targets, head_outputs, anchors):
+        matched_idxs = []
+        for anchors_per_image, targets_per_image in zip(anchors, targets):
+            matched_idxs.append(self.proposal_matcher(targets_per_image["boxes"], anchors_per_image))
+
+        return self.head.compute_loss(targets, head_outputs, anchors, matched_idxs)
+
+    def postprocess_detections(self, class_logits, box_regression, anchors, image_shapes):
+        # type: (Tensor, Tensor, List[Tensor], List[Tuple[int, int]])
+        # TODO: Merge this with roi_heads.RoIHeads.postprocess_detections ?
+        device = class_logits.device
+        num_classes = class_logits.shape[-1]
+
+        scores = torch.sigmoid(class_logits)
+
+        # create labels for each score
+        # the +1 is to make the labels identical to other object detection algorithms that treat background as label 0
+        labels = torch.arange(num_classes, device=device) + 1
+        labels = labels.view(1, -1).expand_as(scores)
+
+        detections = []
+
+        for box_regression_per_image, scores_per_image, labels_per_image, anchors_per_image, image_shape in zip(box_regression, scores, labels, anchors, image_shapes):
+            boxes_per_image = self.box_coder.decode_single(box_regression_per_image, anchors_per_image)
+            boxes_per_image = box_ops.clip_boxes_to_image(boxes_per_image, image_shape)
+
+            image_boxes = []
+            image_scores = []
+            image_labels = []
+
+            for class_index in range(num_classes):
+                # remove low scoring boxes
+                inds = torch.nonzero(scores_per_image[:, class_index] > self.score_thresh).squeeze(1)
+                boxes_per_class, scores_per_class, labels_per_class = boxes_per_image[inds], scores_per_image[inds, class_index], labels_per_image[inds, class_index]
+
+                # remove empty boxes
+                keep = box_ops.remove_small_boxes(boxes_per_class, min_size=1e-2)
+                boxes_per_class, scores_per_class, labels_per_class = boxes_per_class[keep], scores_per_class[keep], labels_per_class[keep]
+
+                # non-maximum suppression, independently done per class
+                keep = box_ops.nms(boxes_per_class, scores_per_class, self.nms_thresh)
+
+                # keep only topk scoring predictions
+                keep = keep[:self.detections_per_img]
+                boxes_per_class, scores_per_class, labels_per_class = boxes_per_class[keep], scores_per_class[keep], labels_per_class[keep]
+
+                image_boxes.append(boxes_per_class)
+                image_scores.append(scores_per_class)
+                image_labels.append(labels_per_class)
```

Comment:
I wonder if we can't optimize this out in order to remove this loop over the number of classes.
For that, I would avoid using `torch.nonzero` in the beginning, and use `batched_nms` instead of `nms`.
This is similar to what I did in `rpn.py` and `roi_heads.py`, and the speed-up would be substantial.

Let me know what you think

Reply:
I will look into this :+1:

Reply:
I've looked into this and I think it is possible, but the logic for `RoIHeads` and `retinanet` is different. For example the shape of `box_regression` for FasterRCNN in `postprocess_detections` is `(2000, 364)`, for 2000 detections (1000 per image) and 4 regression values per class (`91 * 4 = 364`); are the images concatenated in this tensor somehow? I was expecting something like `(2, 1000, 91, 4)`.

The output for retinanet is different, it is shaped `(1, 49104, 4)`. Main differences are that the first dimension is the batch size (number of images), and there is only one bbox regression for each anchor (by design from their paper), as opposed to one per class. This makes the logic in `batched_nms` and `decode` different. What do you advise we do here?

Reply:
> are the images concatenated in this tensor somehow?

yes, for faster r-cnn they are concatenated, which is necessary for RoIAlign to work.

I think that we can still do the same approach, but just paying a bit attention with the indices.

Also, a meta-comment: are you sure that we do the NMS over all the feature map levels at once, and not independently per feature map level?

Reply:
> I can write down the optimized version afterwise, but let's move on with the current version first.

So you'd like me to ignore this optimization for now?

> Also, a meta-comment: are you sure that we do the NMS over all the feature map levels at once, and not independently per feature map level?

Yes I'm sure. After the retinanet heads have been run, they are concatenated, so the concept of pyramid levels is lost there. Is this design desired? Or do you prefer to have the outputs in a list to keep the structure of the pyramid?

Reply:
> So you'd like me to ignore this optimization for now?

Yeah, let's leave this out for now (I won't have the time in the next couple of weeks to get you a snippet)

> Is this design desired? Or do you prefer to have the outputs in a list to keep the structure of the pyramid

Concatenating the results is fine. I thought that they did it per pyramid level, but looks like I was wrong